ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -639,6 +654,36 @@ private IndexValue markAsDeleted(StoreKey id, FileSpan fileSpan, MessageInfo inf
         newValue.getOperationTimeInMs(), newValue.getAccountId(), newValue.getContainerId());
   }
 
+  /**
+   * Marks a blob as permanent
+   * @param id the {@link StoreKey} of the blob
+   * @param fileSpan the file span represented by this entry in the log
+   * @param operationTimeMs the time of the update operation
+   * @return the {@link IndexValue} of the ttl update record
+   * @throws StoreException if there is any problem writing the index record
+   */
+  IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, long operationTimeMs) throws StoreException {
+    validateFileSpan(fileSpan, true);
+    IndexValue value = findKey(id);
+    if (value == null) {
+      throw new StoreException("Id " + id + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
+    } else if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+      throw new StoreException("Id " + id + " deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
+    } else if (value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
+      throw new StoreException("TTL of " + id + " already updated in index" + dataDir,
+          StoreErrorCodes.TTL_Already_Updated);
+    }
+    long size = fileSpan.getEndOffset().getOffset() - fileSpan.getStartOffset().getOffset();
+    IndexValue newValue =
+        new IndexValue(value.getSize(), value.getOffset(), value.getFlags(), Utils.Infinite_Time, operationTimeMs,
+            value.getAccountId(), value.getContainerId());
+    newValue.setFlag(IndexValue.Flags.Ttl_Update_Index);
+    newValue.setNewOffset(fileSpan.getStartOffset());
+    newValue.setNewSize(size);
+    addToIndex(new IndexEntry(id, newValue, null), fileSpan);
+    return new IndexValue(newValue.getOffset().getName(), newValue.getBytes(), newValue.getVersion());
```

Comment:
Is there any potential case we need to return `IndexValue `?

Reply:
Not sure yet - `markAsDeleted()` does - so I did the same here. I will know when I write the code for `BlobStore`

Reply:
This is copied from `markAsDeleted()`.

My _guess_ is that it is because `newValue` is the actual reference within the `IndexSegment` and _if_ some code that received the return did _any_ changes, then it would reflect in the `IndexSegment`. Whether that is something we should guard against, I have no set opinion. I'm ok with returning `newValue`